### DISCUSSION TITLE
Integrating Tailwind CSS in Blazor

### DISCUSSION THREAD

**Discussion by giscus** [timdeschryver/timdeschryver.dev#361] [Announcements]
# Integrating Tailwind CSS in Blazor

In this post, you learn how to set up Tailwind CSS in a Blazor application, how to optimize the development experience, and how to minify the CSS file in a production environment.

https://timdeschryver.dev/blog/integrating-tailwind-css-in-blazor

**Comment by khalil2099** [comment]
why is tailwind so complicated ? it is not a simple CSS just like bootstrap ?

**Reply by timdeschryver** [reply]
The article shows how to include tailwind in the most "performant" way. Adding tailwind at compile-time allows it to trim down its bundle by removing unused css.

To keep things as simple as possible you can also import tailwind using a CDN.
For more info about this, see https://tailwindcss.com/docs/installation/play-cdn

```
 <script src="https://cdn.tailwindcss.com"></script>
```

**Comment by Duelbat** [comment]
This is definitely not clear at all. The command creates a second WWWROOT folder ??? And you are not clear about where this command must be run from? The root of the application? The Client app or the server app ?? I am using the new BLAZOR Webapp template for .net 8.

**Reply by timdeschryver** [reply]
Thanks for your feedback @Duelbat .
I've updated article to provide a better experience.

To answer your points directly:

- The command needs to be run from the root the project (the directory containing the `csproj` file)
- I've updated some examples to include a server and WASM application
- The command doesn't create a second `wwwroot` folder, it only adds the generated file to the folder

**Comment by juhinagpure** [comment]
To integrate Tailwind CSS in a Blazor app :

1. Install Tailwind CSS and dependencies using npm.
2. Set up tailwind.config.js and postcss.config.js for configuration.
3. Modify wwwroot/css/app.css to include Tailwind's directives (@tailwind base, @tailwind components, @tailwind utilities).
4. Use PurgeCSS in the Tailwind config to remove unused CSS in production.
5. Minify the CSS for production for better performance.

**Comment by m-GDEV** [comment]
Good article! I managed to get it working with a similar configuration.

Often I will create components that have dynamic properties like `MaxWidth`.

Then in the component I'll have `<div class="bg-black max-w-[@MaxWidth]></div>`

I'll use the component like so: `<Component MaxWidth="40rem" />

The tailwind CLI does not recognize this style as its computed at runtime.

Have you ever run into this issue? If so, have you found a solution?

Thanks!

**Reply by timdeschryver** [reply]
Thanks, I haven't had the need to do this but I think the docs provide some guidance on this topic:

- https://tailwindcss.com/docs/content-configuration#dynamic-class-names
- https://tailwindcss.com/docs/content-configuration#using-regular-expressions This one in particular.

**Comment by GuyHeylens** [comment]
Hello, thanks for the explanation. I just faced one issue on .Net8. 
In the csproj file you added the 3 targets with the same name. 
In my case this conflicted, and the stylesheet was not generated. 
I solved it using a different name for each of the Targets. 
Kind regards.

**Reply by timdeschryver** [reply]
Thanks for mentioning this issue, I will make sure I'll update the snippet.

**Comment by nhwilly** [comment]
Great article thanks.

**Reply by timdeschryver** [reply]
Thanks @nhwilly !

**Reply by nhwilly** [reply]
Any chance you've tried this on a Blazor WebApp from dotnet 9?  The two projects make it challenging.  I can't seem to get the tailwind.config.js to actually scan the client project.  Since that's what MS is all about with the template, etc.  I thought I'd ask.

**Reply by nhwilly** [reply]
Hey, Tim.  I ended writing my own little blog post so I could remember what to do each time I started a new project.  You might find it useful.

https://www.billtalkstoomuch.com/2025/02/12/installing-tailwind-css-v4-in-a-blazor-webapp/

**Comment by trancikk** [comment]
Great article!

However, Tailwind has been recently updated to v4, it now uses vite plugin or a separate CLI tool.

**Reply by timdeschryver** [reply]
Thanks for pointing that out.

**Comment by djneill** [comment]
Great article, super helpful! I know this was written before Tailwind 4 was released, so I’m curious how your approach might change today.

With TW4 removing the need for a config file in many cases, would you still recommend using a tailwind.config.js in a Blazor project to control content paths and theme extensions? And is there any downside to using the Tailwind CLI directly in TW4, even though it adds node_modules and a package.json where you can add the watch script?

**Reply by timdeschryver** [reply]
Thanks @djneill .
Personally I prefer keeping the config file, just to be explicit about it.
I don't think there are downsides using the CLI, AFAIK any other option also means that you'll have the node_modules folder and a package.json file.
